feat(roadMap): create skeleton roadmap domain entities

SeungYeop-Han opened the pull request:
## Summary
- 로드맵 도메인 엔티티 개발

## Related Issue
- Issue Number: #16 

## Changes
- 모든 엔티티는 대리키를 사용합니다(`IdEntity`)
- `RoadMap`, `Topic`, `SubTopic` 엔티티의 경우 감사(audit) 기능을 사용합니다.(`AggragateRoot`, `IdAuditEntity`)
- `RoadMap` 은 낙관적 락을 사용합니다.(`AggregateRoot`)
- 연관관계의 방향은 전부 양방향으로 설정했습니다.
- 연관관계의 주인은 전부 자식 엔티티입니다.
- 자식 엔티티에 존재하는 부모 엔티티 레퍼런스에 @JsonIgnore 처리하여
JSON 직렬화 시 순환 참조에 의한 무한 루프에 빠지지 않도록 했습니다.
- @OneToMany 관계의 FetchType 은 LAZY 입니다.
- @ManyToOne 관계의 FetchType 은 EAGER 입니다.
- 정적 팩토리 메서드 패턴을 적용했습니다.
- 정적 팩토리 메서드의 파라미터를 통해 전달받은 자식 엔티티 리스트를 순회하며, 각 자식 엔티티의 setXxx 메서드를 호출합니다(양방향 연결)

## What's Next
- 부모 엔티티에 add 및 remove 메서드 생성(순서에 주의하면서)
- 도메인 불변 조건 및 검증 로직 추가
- DB 컬럼 제약 조건 적용
- `@ManyToOne` 컬럼의 `updatable` 여부 고민해보기
- 도메인 에러 코드 정의
- ...

## Discussion Topic
- ...

## Checklist
- [ ] 로컬 테스트 완료
- [x] lint 통과
- [x] 리뷰어 요청(최소 2명), 라벨/스코프 지정
- [x] 브랜치 최신화(rebase) 및 충돌 없음

SeungYeop-Han committed:
refactor(roadmap): Topic 엔티티 클래스의 order 컬럼에 매핑될 db 컬럼 이름을 sort_order 로 변경

order_in_roadmap -> sort_order

#16

SeungYeop-Han committed:
fix(roadmap): Topic 의 SubTopic 으로의 객체 관계 누락 해결

# 16

SeungYeop-Han committed:
refactor(roadmap): 패키지 이름 변경 persistence -> domain

##16

SeungYeop-Han committed:
feat(roadmap): 로드맵 도메인 에러 코드 정의

RoadMapIErrorCode --extends--> IErrorCode

##16

SeungYeop-Han committed:
feat(roadmap): 도메인 엔티티의 정적 팩토리 메서드 파라미터로 사용할 DTO 개발

- 전부 레코드로 개발 진행했습니다.
- 삭제 관련 필드는 포함하지 않았습니다.
- 부모 엔티티도 포함하지 않습니다. 부모 엔티티에서 자식 엔티티의 setXxx 메서드를 통해 양방향 연결합니다.
- DTO 에는 별도의 검증 로직을 추가하지 않았습니다. 나중에 추가할 수도 있습니다.

##16

SeungYeop-Han committed:
feat(roadmap): 도메인 엔티티 개발

- 모든 엔티티는 대리키를 사용합니다(`IdEntity`)
- `RoadMap`, `Topic`, `SubTopic` 엔티티의 경우 감사(audit) 기능을 사용합니다.(`AggragateRoot`, `IdAuditEntity`)
- `RoadMap` 은 낙관적 락을 사용합니다.(`AggregateRoot`)
- 연관관계의 방향은 전부 양방향으로 설정했습니다.
- 연관관계의 주인은 전부 자식 엔티티입니다.
- 자식 엔티티에 존재하는 부모 엔티티 레퍼런스에 @JsonIgnore 처리하여
JSON 직렬화 시 순환 참조에 의한 무한 루프에 빠지지 않도록 했습니다.
- @OneToMany 관계의 FetchType 은 LAZY 입니다.
- @ManyToOne 관계의 FetchType 은 EAGER 입니다.
- 정적 팩토리 메서드 패턴을 적용했습니다.
- 정적 팩토리 메서드의 파라미터를 통해 전달받은 자식 엔티티 리스트를 순회하며, 각 자식 엔티티의 setXxx 메서드를 호출합니다(양방향 연결)

##16

github-actions[bot] committed:
style: apply spotless format automatically

youabledev commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/RoadMap.java:
```diff
@@ -0,0 +1,73 @@
+    private boolean isDraft;
+
+    @Column(name = "category_id", nullable = false)
+    private Long category_id;
```
```suggestion
    private Long categoryId;
```
카멜 케이스로 통일해서 category_id 필드명을 `categoryId` 로 바꿨으면 합니다!

SeungYeop-Han replied:
넵 반영하겠습니다!

winwinhun commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/RoadMap.java:
```diff
@@ -0,0 +1,73 @@
+import com.kllhy.roadmap.common.model.AggregateRoot;
+import com.kllhy.roadmap.roadmap.domain.model.creation_spec.CreationRoadMap;
+import jakarta.persistence.*;
+import java.sql.Timestamp;
```
java.time.LocalDateTime이 아닌 해당코드를 사용하신 이유를 여쭤보고싶습니다!

SeungYeop-Han replied:
사실 팀원 분들이랑 논의했던 부분이기도 했습니다.
관련 Discussion 링크 보내드리겠습니다. 참고 부탁드립니다.
https://github.com/potenup-kllhy/roadmap/discussions/17

dd-jiyun commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/ResourceTopic.java:
```diff
@@ -0,0 +1,60 @@
+    @JoinColumn(name = "topic_id", nullable = false)
+    private Topic topic;
+
+    public ResourceTopic() {}
```
디폴트 생성자를 만들어놓으신 이유가 있으신가요? 

추가로 lombok의 `@NoArgsConstructor` 어노테이션을 사용하셔도 좋을 것 같아요 !

SeungYeop-Han replied:
네 어제 민영님 코드 보면서 그거 사용하면 좋겠다고 생각하긴 했습니다.
피드백 감사드립니다!

dd-jiyun commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/RoadMap.java:
```diff
@@ -0,0 +1,73 @@
+    private boolean isDraft;
+
+    @Column(name = "category_id", nullable = false)
+    private Long category_id;
```
해당 필드명 카멜케이스로 바꾸셔야할 것 같습니다!!

SeungYeop-Han replied:
앗 감사합니다!

winwinhun commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/Topic.java:
```diff
@@ -0,0 +1,108 @@
+    @Column(name = "sort_order", nullable = false)
+    private Integer order;
+
+    @Column(name = "deleted_at", nullable = false)
```
null값을 넣을 수 없는데, line_76의 생성자에서는 null값을 넣으려고 하는 걸로 보입니다. 
이 경우에 jpa가 객체를 insert 할 때 "deleted_at" 필드의 값이 null임을 감지하고, 해당 제약조건에 따라서 DB에 null 값을 전달해서 예외가 발생 할 것 같습니다.
따라서 false를 true로 변경하심이 어떤지 여쭤보고싶습니다!

SeungYeop-Han replied:
피드백 감사합니다. nullable=true 로 두는게 맞는 것 같아요!

dd-jiyun reviewed (commented):
고생하셨습니다 ☺️

minyong96 commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/RoadMap.java:
```diff
@@ -0,0 +1,73 @@
+            cascade = CascadeType.ALL,
+            orphanRemoval = true,
+            fetch = FetchType.LAZY)
+    private List<Topic> topics = new ArrayList<>();
```
불변 리스트로 두는건 어떠신지요?

SeungYeop-Han replied:
혹시 이유를 여쭤봐도 될까요?

minyong96 commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/creation_spec/CreationRoadMap.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.List;
+
+public record CreationRoadMap(
+        String title, String description, boolean isDraft, Long categoryId, List<Topic> topics) {}
```
CreationRoadMap에서 Topic 도메인을 넣고 계신데
이렇게 되면 외부에서 Topic이 생성되어서 불변검증이 어려워지지 않을까요?
엔티티 내부가 아닌 Spec에서 Topic엔티티를 사용하신 이유가 있을까요?

SeungYeop-Han replied:
그거 수정해야 합니다. 저도 어제 집가면서 내일 수정해야 겠다고 생각했던 부분입니다. 피드백 감사합니다!

minyong96 commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/creation_spec/CreationResourceTopic.java:
```diff
@@ -0,0 +1,6 @@
+import com.kllhy.roadmap.roadmap.domain.model.enums.ResourceType;
+
+public record CreationResourceTopic(
+        String name, ResourceType resourceType, Integer order, String link) {}
```
spec을 사용하는 이유가 사실은 1차검증의 목적도 있는데, (엔티티 내부 2차검증)
검증로직을 작성해주면 더 좋을 거 같습니다!

SeungYeop-Han replied:
검증 로직 오늘 추가하겠습니다!

youabledev commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/RoadMap.java:
```diff
@@ -0,0 +1,73 @@
+                        creationSpec.description(),
+                        creationSpec.isDraft(),
+                        creationSpec.categoryId(),
+                        creationSpec.topics());
```
CreationRoadMap creationSpec의 topics가 CreationTopic일 것으로 기대했는데 타입이 Topic이라 의문이 듭니다
위와 같이 되는 경우 CreationRoadMap 객체는 Resource와 SubTopic을 가지고 있는 Topic을 가지고 있어야 합니다

SeungYeop-Han replied:
네 그 부분 수정하겠습니다. 피드백 감사합니다!

youabledev commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/ResourceSubTopic.java:
```diff
@@ -0,0 +1,57 @@
+    @JsonIgnore
+    @ManyToOne(optional = false)
+    @JoinColumn(name = "sub_topic_id", nullable = false)
+    private SubTopic subTopic;
```
제가 작업 하며 놓친 부분 추가해 주셔서 감사합니다!

SeungYeop-Han committed:
fix(roadmap): PR 피드백 반영

- Topic.deletedAt nullable=false -> true(default) 로 변경
- 이름 변경: category_id -> categoryId
- protected 기본 생성자 lombok 으로 구현
- CreationSpec 내에서 엔티티 직접 사용하던 문제 해결
- 정적 팩토리 메서드에 자식 엔티티 리스트 정렬 로직 추가
  - 정렬을 위한 getter 메서드를 lombok 을 이용하여 추가

#16

github-actions[bot] committed:
style: apply spotless format automatically

youabledev reviewed (approved):
고생하셨습니다!

SeungYeop-Han committed:
fix(roadmap): 양방향 연관관계 설정 메서드에서 NPE 를 발생시키도록 변경

Objects.requireNonNull

#16

SeungYeop-Han committed:
feat(roadmap): 엔티티 불변식 조건 검증

CreationSpec DTO 에서 not null 검증
정적 팩토리 메서드 에서는 not null 필드에 대해 null 이 아님을 믿고 검증

#16

github-actions[bot] committed:
style: apply spotless format automatically

youabledev commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/ResourceSubTopic.java:
```diff
@@ -0,0 +1,72 @@
+
+    public static ResourceSubTopic create(CreationResourceSubTopic creationSpec) {
+        String name = creationSpec.name();
+        if (name.isBlank() || name.length() < 2 || 255 < name.length()) {
```
controller에서 사용자에게 데이터 전달 받을 때 validation 으로 꼼꼼히 체크하면 이 단계에서의 추가 검증은 필요 없지 않을까 하는 생각도 있었는데요,
생성 부분에 생성 규칙을 적어 놓으니 하나의 문서가 될 수 있겠다 싶었습니다! 추가하시느라 고생하셨습니다!

youabledev commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/ResourceSubTopic.java:
```diff
@@ -0,0 +1,72 @@
+        String name = creationSpec.name();
+        if (name.isBlank() || name.length() < 2 || 255 < name.length()) {
+            throw new IllegalArgumentException(
+                    "ResourceSubTopic.create: name 이 blank 이거나, 길이가 2 미만 또는 255 초과");
```
혹시 Domain Creat Validation Exception 정의 해서 사용하는 건 어떤 가요?
Exception Handler에서 해당 exception 잡아서 미리 약속된 response로 클라이언트에 전달해 줄 수 있지 않을까 싶습니다!

youabledev commented on src/main/java/com/kllhy/roadmap/roadmap/domain/model/RoadMap.java:
```diff
@@ -0,0 +1,107 @@
+
+        List<Topic> createdTopics =
+                creationSpec.creationTopics() == null
+                        ? Collections.emptyList()
```
저희 정책 정할 때 RoadMap에 적어도 하나 이상의 Topic이 있어야 했던거 같습니다!

<img width="303" height="177" alt="Image" src="https://github.com/user-attachments/assets/38cf8a51-30de-4637-bb6c-ebefaffdacb9" />

이 부분이 맞다면 creationTopics가 하나도 없거나 빈 배열일 때 exception 일으키는게 좋을 것 같은데 어떠신가요?

SeungYeop-Han replied:
제가 이 부분 간과했었네요, 수정하겠습니다!

SeungYeop-Han committed:
fix(roadmap): RoadMap 은 최소 1개의 주제를 포함해야 한다는 정책 누락 해결

#16

github-actions[bot] committed:
style: apply spotless format automatically
